drivers/usb/device/usb_dc_dw.c

```diff
@@ -369,142 +384,201 @@ static int usb_dw_ep_set(u8_t ep,
+		return;
 	}
 
-	if (data_len > avail_space) {
-		data_len = avail_space;
+	/* How much data remaining */
+	to_write = min(ep_ctrl->mps, ep_ctrl->transfer_rem);
+
+	/* Write Data to FIFO */
+	usb_dw_write_fifo(USB_DW_EP_ADDR2IDX(ep), ep_ctrl->transfer_buf,
+			  to_write);
+
+	/* Update Transfer Info */
+	ep_ctrl->transfer_buf += to_write;
+	ep_ctrl->transfer_rem -= to_write;
+	ep_ctrl->transfer_size += to_write;
+}
+
+static inline void usb_dw_fout(u8_t ep, size_t bcnt)
+{
+	struct usb_ep_ctrl_prv *ep_ctrl = USB_DW_EP_ADDR2PRV(ep);
+
+	/* Any Ongoing Transfer ? */
+	if (!ep_ctrl->transfer_buf) {
+		return;
 	}
 
-	if (data_len != 0) {
-		/* Get max packet size and packet count for ep */
-		if (ep_idx == USB_DW_IN_EP_0) {
-			max_pkt_cnt =
-			    USB_DW_DIEPTSIZ0_PKT_CNT_MASK >>
-			    USB_DW_DEPTSIZ_PKT_CNT_OFFSET;
-			max_xfer_size =
-			    USB_DW_DEPTSIZ0_XFER_SIZE_MASK >>
-			    USB_DW_DEPTSIZ_XFER_SIZE_OFFSET;
+	/* Read available data */
+	usb_dw_read_fifo(USB_DW_EP_ADDR2IDX(ep), ep_ctrl->transfer_buf, bcnt);
+
+	/* Update Transfer Info */
+	ep_ctrl->transfer_buf += bcnt;
+	ep_ctrl->transfer_rem -= bcnt;
+	ep_ctrl->transfer_size += bcnt;
+}
+
+/* for ep_read/ep_write compatibility */
+static void usb_dw_transfer_cb_legacy(u8_t ep, int status, size_t tsize)
+{
+	struct usb_ep_ctrl_prv *ep_ctrl = USB_DW_EP_ADDR2PRV(ep);
+
+	ARG_UNUSED(status);
+	ARG_UNUSED(tsize);
+
+	if (ep_ctrl->cb) {
+		if (USB_DW_EP_ADDR2DIR(ep) == USB_EP_DIR_OUT) {
+			ep_ctrl->cb(ep, USB_DC_EP_DATA_OUT);
 		} else {
-			max_pkt_cnt =
-			    USB_DW_DIEPTSIZn_PKT_CNT_MASK >>
-			    USB_DW_DEPTSIZ_PKT_CNT_OFFSET;
-			max_xfer_size =
-			    USB_DW_DEPTSIZn_XFER_SIZE_MASK >>
-			    USB_DW_DEPTSIZ_XFER_SIZE_OFFSET;
+			ep_ctrl->cb(ep, USB_DC_EP_DATA_IN);
 		}
+	}
+}
 
-		/* Check if transfer len is too big */
-		if (data_len > max_xfer_size) {
-			SYS_LOG_WRN("USB IN EP%d len too big (%d->%d)", ep_idx,
-			    data_len, max_xfer_size);
-			data_len = max_xfer_size;
-		}
+static inline void usb_dw_transfer_complete(u8_t ep)
+{
+	struct usb_ep_ctrl_prv *ep_ctrl = USB_DW_EP_ADDR2PRV(ep);
 
-		/*
-		 * Program the transfer size and packet count as follows:
-		 *
-		 * transfer size = N * ep_maxpacket + short_packet
-		 * pktcnt = N + (short_packet exist ? 1 : 0)
-		 */
+	/* Any Ongoing Transfer ? */
+	if (!ep_ctrl->transfer_buf) {
+		return;
+	}
 
-		pkt_cnt = (data_len + ep_mps - 1) / ep_mps;
+	ep_ctrl->transfer_buf = NULL;
+	ep_ctrl->transfer_result = 0;
+	k_sem_give(&ep_ctrl->transfer_sem);
 
-		if (pkt_cnt > max_pkt_cnt) {
-			SYS_LOG_WRN("USB IN EP%d pkt count too big (%d->%d)",
-			    ep_idx, pkt_cnt, pkt_cnt);
-			pkt_cnt = max_pkt_cnt;
-			data_len = pkt_cnt * ep_mps;
-		}
-	} else {
-		/* Zero length packet */
+	/* Asynchronous transfer callback */
+	if (ep_ctrl->transfer_cb) {
+		ep_ctrl->transfer_cb(ep, 0, ep_ctrl->transfer_size);
+	}
+}
+
+static int usb_dc_ep_transfer(const u8_t ep, u8_t *buf, size_t dlen,
+			      bool is_in, usb_dc_transfer_callback cb)
```

Comment:
I see that usb_dc_ep_transfer() is used only in usb_dc_ep_write(). Are you going to export it to API?
Using it inside usb_dc_ep_write() changes a bit logic. Now with every packet send I get warning with EAGAIN error returned, my USB try_write() helper handles this and another attempt succeeds. You get this warning only when other debug disabled, timing does matter.

Reply:
Hi @finikorg, you're correct, plan is to export this API, but before, I would like to have the implementation enabled in both drivers (dw, stm). Then I think we could use this method in all existing class drivers progressively and maybe remove the current usb_read/write API.

I agree that this change a bit the behavior by returning EAGAIN when transfer is ongoing, but I think this is aligned with the original usb_write behavior. Documentation says: 

"To transmit data to the host, the class driver should call usb_write(). Upon completion the registered endpoint callback will be called. Before sending another packet the class driver should wait for the completion of the previous transfer."

In current usb_write implementation we've workaround-ed/hacked a bit the behavior to actively wait for the FIFO availability which basically works but is not the cleanest way.

I'm not sure about the right way to address your comment, if this is OK for you, I propose to add a patch in this PR removing the warning message (causing delays) since EAGAIN is expected when using ep_write without waiting for the ep_callback. Let me know...

Reply:
@finikorg, In order to avoid this regression, I updated usb_dc_ep_write to keep the legacy behavior (active waiting). let me know if it's ok for you now.

Reply:
Now it looks better that we keep legacy logic, when you finish the API we might moved to the new one.